src/core/varlink-cgroup.c

```diff
@@ -0,0 +1,342 @@
+                                &v,
+                                SD_JSON_BUILD_PAIR_STRING("source", nft_set_source_to_string(nft_set->source)),
+                                SD_JSON_BUILD_PAIR_STRING("protocol", nfproto_to_string(nft_set->nfproto)),
+                                SD_JSON_BUILD_PAIR_STRING("table", nft_set->table),
+                                SD_JSON_BUILD_PAIR_STRING("set", nft_set->set));
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+static int bpf_program_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        CGroupBPFForeignProgram *programs = userdata;
+        int r;
+
+        assert(ret);
+        assert(name);
+
+        LIST_FOREACH(programs, p, programs) {
+                r = sd_json_variant_append_arraybo(
+                                &v,
+                                SD_JSON_BUILD_PAIR_STRING("attachType", bpf_cgroup_attach_type_to_string(p->attach_type)),
+                                SD_JSON_BUILD_PAIR_STRING("path", p->bpffs_path));
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+
+static int _device_allow_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        CGroupDeviceAllow *allow = userdata;
+        int r;
+
+        LIST_FOREACH(device_allow, a, allow) {
+                r = sd_json_variant_append_arraybo(
+                                &v,
+                                SD_JSON_BUILD_PAIR_STRING("path", a->path),
+                                SD_JSON_BUILD_PAIR_STRING("permissions", cgroup_device_permissions_to_string(a->permissions)));
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+
+static int controllers_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        CGroupMask *mask = ASSERT_PTR(userdata);
+        int r;
+
+        assert(ret);
+        assert(name);
+
+        for (CGroupController ctrl = 0; ctrl < _CGROUP_CONTROLLER_MAX; ctrl++) {
+                if ((*mask & CGROUP_CONTROLLER_TO_MASK(ctrl)) == 0)
+                        continue;
+
+                r = sd_json_variant_append_arrayb(&v, SD_JSON_BUILD_STRING(cgroup_controller_to_string(ctrl)));
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+
+int unit_cgroup_context_build_json(sd_json_variant **ret, const char *name, void *userdata) {
+        Unit *u = ASSERT_PTR(userdata);
+
+        assert(ret);
+        assert(name);
+
+        CGroupContext *c = unit_get_cgroup_context(u);
```

Comment:
As mentioned, this should take CGContext directly. Please move this call to caller